File: Packages/com.wevaui/Runtime/Css/Cascade/KeywordResolver.cs
using System;
using System.Collections.Generic;

namespace Weva.Css.Cascade {
    internal static class KeywordResolver {
        public static bool IsCssWideKeyword(string value) {
            if (value == null) return false;
            return Match(value, "inherit") || Match(value, "initial") || Match(value, "unset") || Match(value, "revert") || Match(value, "revert-layer");
        }

        // CSS Cascade L4 §7.5 — CSS-wide keyword resolution. The `revert` /
        // `revert-layer` rollback resolution requires the full per-element
        // match list and is performed by the caller via PreResolveRollback
        // BEFORE calling this; once rollback has substituted the winning
        // value, only the four "plain" keywords survive into Resolve.
        // If `revert` / `revert-layer` reach Resolve, the rollback found no
        // lower-priority match (no UA/user-origin rule, no lower layer) and
        // the spec collapses them to `initial`.
        public static string Resolve(string property, string value, ComputedStyle parent) {
            if (value == null) return null;
            string trimmed = value.Trim();
            if (Match(trimmed, "inherit")) return ResolveInherit(property, parent);
            if (Match(trimmed, "initial")) return CssProperties.InitialValueOf(property);
            if (Match(trimmed, "unset")) {
                if (CssProperties.IsInherited(property)) return ResolveInherit(property, parent);
                return CssProperties.InitialValueOf(property);
            }
            if (Match(trimmed, "revert")) return CssProperties.InitialValueOf(property);
            if (Match(trimmed, "revert-layer")) return CssProperties.InitialValueOf(property);
            return value;
        }

        // CSS Cascade L5 §7.4 / §7.5 — rollback resolution. Substitutes the
        // value text of the appropriate lower-priority match for the property
        // when the winner's value is `revert` or `revert-layer`. Returns the
        // rolled-back value text (which may itself be `inherit` / `initial` /
        // even another `revert` for chains) so the caller can re-run the
        // standard CSS-wide keyword resolver on the result.
        //
        // Returns null when no rollback target exists — caller leaves the
        // original `revert` / `revert-layer` token in place, and Resolve()
        // above maps it to initial per spec.
        //
        // `expanded` is the full per-element match list in cascade order
        // (last entry for each property is the winner). `winner` is the
        // current winning match for `property`.
        public static string PreResolveRollback(string property, string value,
                List<MatchedDeclaration> expanded, MatchedDeclaration winner) {
            if (value == null) return null;
            // Tight loop on chained `revert` / `revert-layer` (the UA value
            // might itself be `revert`, which collapses to the next-lower
            // origin, etc.). Capped at 4 hops — deeper chains are pathological.
            string current = value;
            MatchedDeclaration currentWinner = winner;
            for (int depth = 0; depth < 4; depth++) {
                if (current == null) return null;
                string trimmed = current.Trim();
                if (Match(trimmed, "revert")) {
                    var rb = RollbackRevert(property, expanded, currentWinner);
                    if (rb.Value == null) return current; // initial fallback applied downstream
                    current = rb.Value;
                    currentWinner = rb.Match;
                    continue;
                }
                if (Match(trimmed, "revert-layer")) {
                    var rb = RollbackRevertLayer(property, expanded, currentWinner);
                    if (rb.Value == null) return current;
                    current = rb.Value;
                    currentWinner = rb.Match;
                    continue;
                }
                return current;
            }
            return current;
        }

        // `revert`: roll back to the latest match at an origin BELOW the
        // current winner's origin (Author > User > UA in normal order). The
        // !important inversion is intentionally NOT honoured in v1 — the
        // pinned tests in CssWideKeywordTests cover normal-priority cases
        // only.
        static (string Value, MatchedDeclaration Match) RollbackRevert(string property,
                List<MatchedDeclaration> expanded, MatchedDeclaration winner) {
            DeclarationOrigin winnerOrigin = winner.Origin;
            // Latest match (i.e. highest cascade priority) at any origin
            // strictly below winnerOrigin. Scan backwards so the first hit
            // is the answer.
            for (int i = expanded.Count - 1; i >= 0; i--) {
                var m = expanded[i];
                if (m.Declaration.Property != property) continue;
                if (m.Origin >= winnerOrigin) continue;
                return (m.Declaration.ValueText, m);
            }
            return (null, default);
        }

        // `revert-layer`: roll back to the latest match at the SAME origin
        // but at a layer with a strictly lower ordinal. The highest such
        // layer's latest match wins; if none, fall through to revert
        // behaviour (drop the origin entirely).
        static (string Value, MatchedDeclaration Match) RollbackRevertLayer(string property,
                List<MatchedDeclaration> expanded, MatchedDeclaration winner) {
            DeclarationOrigin winnerOrigin = winner.Origin;
            int winnerLayer = winner.LayerOrdinal;
            // Pass 1 — find the largest layer ordinal below winnerLayer at winnerOrigin.
            int maxLowerLayer = -1;
            bool found = false;
            for (int i = 0; i < expanded.Count; i++) {
                var m = expanded[i];
                if (m.Declaration.Property != property) continue;
                if (m.Origin != winnerOrigin) continue;
                if (m.LayerOrdinal >= winnerLayer) continue;
                if (!found || m.LayerOrdinal > maxLowerLayer) {
                    maxLowerLayer = m.LayerOrdinal;
                    found = true;
                }
            }
            if (!found) {
                // No lower-layer match at this origin → fall through to revert.
                return RollbackRevert(property, expanded, winner);
            }
            // Pass 2 — latest match (highest cascade priority) at (winnerOrigin, maxLowerLayer).
            MatchedDeclaration best = default;
            string value = null;
            for (int i = 0; i < expanded.Count; i++) {
                var m = expanded[i];
                if (m.Declaration.Property != property) continue;
                if (m.Origin != winnerOrigin) continue;
                if (m.LayerOrdinal != maxLowerLayer) continue;
                best = m;
                value = m.Declaration.ValueText;
            }
            return (value, best);
        }

        static string ResolveInherit(string property, ComputedStyle parent) {
            if (parent != null && parent.TryGet(property, out var pv)) return pv;
            return CssProperties.InitialValueOf(property);
        }

        static bool Match(string a, string b) {
            return a.Equals(b, StringComparison.OrdinalIgnoreCase);
        }
    }
}

File: libweva/src/keyword_resolver.cpp
#include "weva/keyword_resolver.h"

#include "weva/css_properties.h"

namespace weva {

namespace {

bool iequals(std::string_view a, std::string_view b) {
    if (a.size() != b.size()) return false;
    for (size_t i = 0; i < a.size(); ++i) {
        char x = a[i], y = b[i];
        if (x >= 'A' && x <= 'Z') x = static_cast<char>(x - 'A' + 'a');
        if (y >= 'A' && y <= 'Z') y = static_cast<char>(y - 'A' + 'a');
        if (x != y) return false;
    }
    return true;
}

std::string_view trim(std::string_view s) {
    const auto ws = [](char c) {
        return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f';
    };
    size_t b = 0, e = s.size();
    while (b < e && ws(s[b])) ++b;
    while (e > b && ws(s[e - 1])) --e;
    return s.substr(b, e - b);
}

// The parent's COMPUTED value. This port resolves inheritance lazily, so the
// parent's get() already walks its own chain and falls back to the initial
// value — which is exactly what the C# reads off a materialised parent style.
std::string inherited_value(int property_id, const ComputedStyle* parent) {
    if (parent) return std::string(parent->get(property_id));
    return std::string(CssPropertyRegistry::instance().initial_value(property_id));
}

// Finds the value text of the highest-priority match for `property` at an
// origin strictly below the winner's. `expanded` is in ascending cascade order,
// so scanning backwards makes the first hit the answer.
const MatchedDeclaration* rollback_revert(std::string_view property,
                                          const std::vector<MatchedDeclaration>& expanded,
                                          DeclarationOrigin winner_origin) {
    for (size_t i = expanded.size(); i-- > 0;) {
        const MatchedDeclaration& m = expanded[i];
        if (m.declaration->property != property) continue;
        if (static_cast<int>(m.origin) >= static_cast<int>(winner_origin)) continue;
        return &m;
    }
    return nullptr;
}

// `revert-layer` rolls back to the same origin but the nearest lower layer;
// with no lower layer at that origin it degrades to `revert` and drops the
// origin entirely.
const MatchedDeclaration* rollback_revert_layer(std::string_view property,
                                                const std::vector<MatchedDeclaration>& expanded,
                                                DeclarationOrigin winner_origin,
                                                int winner_layer) {
    int max_lower = -1;
    bool found = false;
    for (const MatchedDeclaration& m : expanded) {
        if (m.declaration->property != property) continue;
        if (m.origin != winner_origin) continue;
        if (m.layer_ordinal >= winner_layer) continue;
        if (!found || m.layer_ordinal > max_lower) {
            max_lower = m.layer_ordinal;
            found = true;
        }
    }
    if (!found) return rollback_revert(property, expanded, winner_origin);

    const MatchedDeclaration* best = nullptr;
    for (const MatchedDeclaration& m : expanded) {
        if (m.declaration->property != property) continue;
        if (m.origin != winner_origin) continue;
        if (m.layer_ordinal != max_lower) continue;
        best = &m;   // ascending cascade order, so the last hit is the winner
    }
    return best;
}

} // namespace

bool is_css_wide_keyword(std::string_view value) {
    // Length-gated so the common case — an ordinary value — costs one compare.
    if (value.size() < 5 || value.size() > 12) return false;
    return iequals(value, "inherit") || iequals(value, "initial") ||
           iequals(value, "unset") || iequals(value, "revert") ||
           iequals(value, "revert-layer");
}

bool resolve_css_wide_keyword(int property_id, std::string_view value,
                              const ComputedStyle* parent, std::string* out) {
    const std::string_view t = trim(value);
    if (!is_css_wide_keyword(t)) return false;
    auto& reg = CssPropertyRegistry::instance();

    if (iequals(t, "inherit")) {
        *out = inherited_value(property_id, parent);
    } else if (iequals(t, "unset")) {
        *out = reg.is_inherited(property_id) ? inherited_value(property_id, parent)
                                             : std::string(reg.initial_value(property_id));
    } else {
        // initial, revert and revert-layer. The two rollback keywords only
        // reach here when no rollback target existed.
        *out = std::string(reg.initial_value(property_id));
    }
    return true;
}

bool resolve_css_wide_keyword_custom(std::string_view name, std::string_view value,
                                     const ComputedStyle* parent, std::string* out) {
    const std::string_view t = trim(value);
    if (!is_css_wide_keyword(t)) return false;

    // `inherit` and `unset` coincide: every custom property is inherited.
    if (iequals(t, "inherit") || iequals(t, "unset")) {
        if (parent && parent->contains(name)) {
            *out = std::string(parent->get(name));
        } else {
            out->clear();
        }
        return true;
    }
    // The initial value of a custom property is the empty string, even for one
    // registered via `@property` — the C# reads the initial value off a
    // synthesised descriptor that knows nothing about the registry.
    out->clear();
    return true;
}

std::string pre_resolve_rollback(std::string_view property, std::string_view value,
                                 const std::vector<MatchedDeclaration>& expanded,
                                 const CascadeKey& winner) {
    std::string current(value);
    DeclarationOrigin origin = winner.origin;
    int layer = winner.layer_ordinal;

    // A rolled-back value may itself be `revert`, which rolls back again.
    // Capped at four hops; deeper chains are pathological.
    for (int depth = 0; depth < 4; ++depth) {
        const std::string_view t = trim(current);
        const bool is_revert = iequals(t, "revert");
        const bool is_revert_layer = iequals(t, "revert-layer");
        if (!is_revert && !is_revert_layer) return current;

        const MatchedDeclaration* rb =
            is_revert ? rollback_revert(property, expanded, origin)
                      : rollback_revert_layer(property, expanded, origin, layer);
        // No target: leave the keyword in place so the resolver maps it to
        // the initial value.
        if (!rb) return current;
        current = rb->declaration->value_text;
        origin = rb->origin;
        layer = rb->layer_ordinal;
    }
    return current;
}

} // namespace weva
